Add forward extremities endpoint to rooms admin API (#9062)
matrix-org/synapse: docs/admin_api/rooms.md

        "user_id": "@foo:example.com"
    }
```

# Forward Extremities Admin API

Enables querying and deleting forward extremities from rooms. When a lot of forward

> extremities accumulate in a room, performance can become degraded.

Reviewer: Is there a reference issue for this or anything that we should link to?

Author: Will add a link to https://github.com/matrix-org/synapse/issues/1760

When using this API endpoint to delete any extra forward extremities for a room, 
the server does not need to be restarted as the relevant caches will be cleared 

> in the API call.

Reviewer: I don't think this is necessary to include here? This seems more developer focused than operator focused.

## Check for forward extremities

> To check the status of forward extremities for a room:

Reviewer: What sort of situations would you use this in? Is the state group useful?

Author: No, probably not particularly, except maybe for developers. I did think of returning just a count, but it felt that since the information "is there", it felt unnecessary to not return results as well.

My understanding of the needs for this API endpoint from an operational point of view is mainly to see the count, though I would maybe check with the EMS ops team first to confirm.

Reviewer: I suspect that's just internal state then and not worth returning. @erikjohnston do you see any value in returning those?

Reviewer 2: I've sometimes used the state group as a proxy for how big the state difference was between extremities were, but it's a bad heuristic. I think a more useful one would probably be `depth` and `received_ts`, as that'll give you a better way of looking for stuck extremities

Author: Do we consider those to be useful in this admin API? I guess the options are:

1) keep as is
2) remove `state_group`
3) remove `state_group`, add `depth` and `received_ts`
4) remove everything, just return a count
5) something else? :)

I don't have enough experience on the extremities to vote really.

Reviewer 2: I think adding `depth` and `received_ts`  is probably the right thing

```
    GET /_synapse/admin/v1/rooms/<room_id_or_alias>/forward_extremities
```

A response as follows will be returned:

```json
{
  "count": 1,
  "results": [
    {
      "event_id": "$M5SP266vsnxctfwFgFLNceaCo3ujhRtg_NiiHabcdfgh",
      "state_group": 439
    }
  ]
}    
```

## Deleting forward extremities

> In the event a room has lots of forward extremities, the extra can be
> deleted as follows:

Reviewer: "In the event" is a bit confusing when we're discussing an API that deals with events. 😄 

```suggestion
If a room has lots of forward extremities, the extra can be
deleted as follows:
```

```
    DELETE /_synapse/admin/v1/rooms/<room_id_or_alias>/forward_extremities
```

A response as follows will be returned, indicating the amount of forward extremities
that were deleted.

```json
{
  "deleted": 1
}
```

The cache `get_latest_event_ids_in_room` will be invalidated, if any forward extremities

> were deleted.

Reviewer: I don't think it is worth mentioning this in the documentation. By the nature of the API existing it should be implied that it is doing the Right Thing™️.
